File: Arrays & Hashing/560. Subarray Sum Equals K/Solution.py
from collections import defaultdict
from typing import List


class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        total = 0
        currSum = 0
        pre = defaultdict(int)  # noqa: F821
        pre[0] = 1

        for num in nums:
            currSum += num
            if currSum - k in pre:
                total += pre[currSum - k]
            pre[currSum] += 1
        return total

File: Arrays & Hashing/560. Subarray Sum Equals K/Solution.java
import java.util.HashMap;

class Solution {
    public int subarraySum(int[] nums, int k) {
        int total = 0;
        int currSum = 0;
        HashMap<Integer, Integer> pre = new HashMap<>();
        pre.put(0, 1);
        
        for (int num: nums){
            currSum += num;
            total += pre.getOrDefault(currSum - k, 0);
            pre.put(currSum, pre.getOrDefault(currSum, 0) + 1);
        }
        return total;
    }
}